src/map.rs

```diff
@@ -1020,15 +1020,13 @@ where
                         // the new bin will also be a tree bin. if both the high
                         // bin and the low bin are non-empty, we have to
                         // allocate a new TreeBin.
-                        Shared::boxed(
-                            BinEntry::Tree(TreeBin::new(
-                                // safety: we have just created `low` and its `next`
-                                // nodes and have never shared them
-                                unsafe { low.into_box() },
-                                guard,
-                            )),
-                            &self.collector,
-                        )
+                        // safety: we have just created `low` and its `next`
+                        // nodes and have never shared them
```

Comment:
I'd like us to update this (and the other) safety comment to also reflect the new safety requirement of `TreeBin::new` (namely that the pointer in the first argument was created by a `Box`).

Reply:
I reworked `TreeBin::new` and the safety comments around `Shared::boxed` to avoid the `Shared->*mut T->Shared` conversions entirely.